File: run_pipeline.py
"""
run_pipeline.py  —  one entry point for the whole daily run.

Runs each stage in order:  apple fetch  ->  spotify fetch  ->  youtube fetch  ->  loader

Each stage runs as its OWN subprocess (a separate file = a separate process). That isolation
is the point: if a stage errors, hangs, or even crashes hard, we kill/abandon just that one
process, log what happened, and move on to the next stage. One platform's outage never stops
the others, and the loader still runs at the end — it simply loads whatever files exist and
keeps yesterday's data for whatever's missing.

Why the loader runs even if fetches failed: the loader treats a missing/broken file as a skip,
so a failed Spotify fetch just means "serve yesterday's Spotify" while Apple + YouTube update.

Edit STAGES below to match your actual fetch-script filenames and give each a sane timeout.
"""

from __future__ import annotations

import logging
import subprocess
import sys
from datetime import date, datetime
from pathlib import Path
from dotenv import load_dotenv
load_dotenv()

BASE_DIR = Path(__file__).resolve().parent
LOG_DIR = BASE_DIR / "logs"
RUN_DATE = date.today()

# (label, script filename, timeout in seconds).  Order matters: fetches first, loader last.
# Timeouts are generous because Apple alone is ~175 countries x genres; tune to your machine.
STAGES = [
    ("apple fetch",   "scrapers.apple_test",   7200),
    ("spotify fetch", "scrapers.spotify_test", 3600),
    ("youtube fetch", "scrapers.youtube_test", 1800),
    ("loader",        "loader.py",             1800),
]


def setup_logging() -> logging.Logger:
    LOG_DIR.mkdir(parents=True, exist_ok=True)
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s  %(levelname)-7s  %(message)s",
        handlers=[
            logging.StreamHandler(sys.stdout),
            logging.FileHandler(LOG_DIR / f"pipeline_{RUN_DATE}.log", encoding="utf-8"),
        ],
    )
    return logging.getLogger("pipeline")


log = setup_logging()


def run_stage(label: str, script: str, timeout: int) -> tuple[str, str]:
    """Run one stage in isolation. Returns (label, status). Never raises."""
    is_module = not script.endswith(".py")
    if not is_module:
        path = BASE_DIR / script
        if not path.exists():
            log.error("[%s] script not found: %s — skipping stage", label, path)
            return label, "missing"

    log.info("[%s] starting  (%s, timeout %ds)", label, script, timeout)
    start = datetime.now()
    try:
        if is_module:
            cmd = [sys.executable, "-m", script]
        else:
            cmd = [sys.executable, str(path)]
        result = subprocess.run(
            cmd,
            cwd=str(BASE_DIR),
            capture_output=True,
            text=True,
            timeout=timeout,
        )
    except subprocess.TimeoutExpired:
        # subprocess.run has already killed the child process for us.
        log.error("[%s] TIMED OUT after %ds — process killed, continuing", label, timeout)
        return label, "timeout"
    except Exception as exc:  # noqa: BLE001 - never let a launch error stop the pipeline
        log.error("[%s] could not run (%s) — continuing", label, exc)
        return label, "error"

    elapsed = (datetime.now() - start).total_seconds()

    # Pipe the child's output into our log so everything lives in one place.
    if result.stdout:
        log.info("[%s] stdout (tail):\n%s", label, _tail(result.stdout))
    if result.returncode != 0:
        log.error("[%s] FAILED rc=%d after %.0fs — continuing", label, result.returncode, elapsed)
        if result.stderr:
            log.error("[%s] stderr (tail):\n%s", label, _tail(result.stderr))
        return label, "failed"

    log.info("[%s] done  rc=0  %.0fs", label, elapsed)
    return label, "ok"


def _tail(text: str, lines: int = 25) -> str:
    return "\n".join(text.strip().splitlines()[-lines:])


def main() -> int:
    log.info("=" * 70)
    log.info("pipeline start  run_date=%s  dir=%s", RUN_DATE, BASE_DIR)

    results = [run_stage(label, script, timeout) for label, script, timeout in STAGES]

    log.info("-" * 70)
    log.info("pipeline summary:")
    for label, status in results:
        log.info("    %-14s %s", label, status.upper())

    failures = [label for label, status in results if status != "ok"]
    log.info("pipeline done  %d/%d stages ok", len(results) - len(failures), len(results))
    log.info("=" * 70)

    # Exit code = number of stages that didn't succeed, so cron/monitoring can alarm
    # while the pipeline itself still ran every stage to completion.
    return len(failures)


if __name__ == "__main__":
    raise SystemExit(main())

File: run_pipeline.php
<?php

namespace App\Scripts;

$baseDir = __DIR__;
$logDir = realpath($baseDir . '/../logs');
if (!$logDir) {
    mkdir($baseDir . '/../logs', 0777, true);
    $logDir = realpath($baseDir . '/../logs');
}

$runDate = date('Y-m-d');
$logFile = $logDir . "/pipeline_{$runDate}.log";

// 1. Define the stages and their order
$stages = [
    ['label' => 'apple fetch',   'script' => 'scrapers/apple_test.php'],
    ['label' => 'spotify fetch', 'script' => 'scrapers/spotify_test.php'],
    ['label' => 'youtube fetch', 'script' => 'scrapers/youtube_test.php'],
    ['label' => 'meta gen',      'script' => 'generate_meta.php'],
    ['label' => 'loader',        'script' => 'loader.php'],
];

// --- Helpers ---
function logMsg($level, $message) {
    global $logFile;
    $time = date('Y-m-d H:i:s,000');
    $line = "{$time}  " . str_pad($level, 7) . "  {$message}\n";
    echo $line;
    file_put_contents($logFile, $line, FILE_APPEND);
}

function tail($text, $lines = 25) {
    if (empty(trim($text))) return "";
    $linesArr = explode("\n", trim($text));
    $tailArr = array_slice($linesArr, -$lines);
    return implode("\n", $tailArr);
}

// --- Main Execution ---
function main() {
    global $baseDir, $runDate, $stages;

    logMsg('INFO', str_repeat('=', 70));
    logMsg('INFO', "pipeline start  run_date={$runDate}  dir={$baseDir}");

    $results = [];

    foreach ($stages as $stage) {
        $label = $stage['label'];
        $script = $stage['script'];
        $scriptPath = $baseDir . '/' . $script;

        if (!file_exists($scriptPath)) {
            logMsg('ERROR', "[{$label}] script not found: {$scriptPath} — skipping stage");
            $results[] = ['label' => $label, 'status' => 'missing'];
            continue;
        }

        logMsg('INFO', "[{$label}] starting  ({$script})");
        $start = microtime(true);

        // Execute the script as an isolated sub-process.
        // "2>&1" ensures we capture both standard output and fatal PHP errors (stderr).
        $cmd = "php " . escapeshellarg($scriptPath) . " 2>&1";
        
        // Run the command and capture output
        exec($cmd, $outputArray, $exitCode);
        
        $elapsed = round(microtime(true) - $start);
        $outputText = implode("\n", $outputArray);
        
        // Clear the array for the next loop iteration
        unset($outputArray); 

        if ($outputText) {
            logMsg('INFO', "[{$label}] output (tail):\n" . tail($outputText));
        }

        if ($exitCode !== 0) {
            logMsg('ERROR', "[{$label}] FAILED rc={$exitCode} after {$elapsed}s — continuing");
            $results[] = ['label' => $label, 'status' => 'failed'];
        } else {
            logMsg('INFO', "[{$label}] done  rc=0  {$elapsed}s");
            $results[] = ['label' => $label, 'status' => 'ok'];
        }
    }

    logMsg('INFO', str_repeat('-', 70));
    logMsg('INFO', "pipeline summary:");

    $failures = 0;
    foreach ($results as $res) {
        logMsg('INFO', "    " . str_pad($res['label'], 14) . " " . strtoupper($res['status']));
        if ($res['status'] !== 'ok') {
            $failures++;
        }
    }

    $okCount = count($stages) - $failures;
    $totalCount = count($stages);
    logMsg('INFO', "pipeline done  {$okCount}/{$totalCount} stages ok");
    logMsg('INFO', str_repeat('=', 70));

    // Exit code = number of stages that didn't succeed
    return $failures;
}

exit(main());